# ANN 02: Fashion-MNIST Logistic Regression

이번 챕터에서는 **로지스틱 회귀(Logistic Regression)**를 사용해  
Fashion-MNIST 데이터셋을 분류합니다.

📌 목표:
- Dense(Softmax)만 사용해 간단한 모델 구축
- ANN (chapter01)과 성능 비교


In [1]:
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.font_manager as fm
import platform
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

# 한글 폰트 설정
if platform.system() == 'Linux':
    plt.rcParams['font.family'] = 'NanumGothic'
elif platform.system() == 'Darwin':
    plt.rcParams['font.family'] = 'AppleGothic'
else:
    plt.rcParams['font.family'] = 'Malgun Gothic'

plt.rcParams['axes.unicode_minus'] = False

print("✅ TensorFlow 버전:", tf.__version__)


2025-09-01 17:55:59.530848: I external/local_xla/xla/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
2025-09-01 17:55:59.647568: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2025-09-01 17:56:01.691196: I external/local_xla/xla/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.


✅ TensorFlow 버전: 2.20.0


In [2]:
(x_train, y_train), (x_test, y_test) = keras.datasets.fashion_mnist.load_data()

# 정규화 (0~1 범위)
x_train = x_train / 255.0
x_test = x_test / 255.0

print("학습 데이터:", x_train.shape, y_train.shape)
print("테스트 데이터:", x_test.shape, y_test.shape)


학습 데이터: (60000, 28, 28) (60000,)
테스트 데이터: (10000, 28, 28) (10000,)


## 🧮 Logistic Regression 모델 구조
- Flatten: 28x28 → 784 벡터
- Dense(10, softmax): 클래스 10개 분류


In [3]:
model = keras.Sequential([
    layers.Flatten(input_shape=(28, 28)),
    layers.Dense(10, activation="softmax")
])

model.summary()


/home/ccd/dl/v_dl/lib/python3.10/site-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)
2025-09-01 17:56:03.334604: E external/local_xla/xla/stream_executor/cuda/cuda_platform.cc:51] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten (Flatten)               │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 10)             │         7,850 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 7,850 (30.66 KB)

 Trainable params: 7,850 (30.66 KB)

 Non-trainable params: 0 (0.00 B)

In [4]:
model.compile(optimizer="adam",
              loss="sparse_categorical_crossentropy",
              metrics=["accuracy"])

history = model.fit(x_train, y_train,
                    epochs=5,
                    validation_split=0.1,
                    batch_size=32)


Epoch 1/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7929 - loss: 0.6149 - val_accuracy: 0.8313 - val_loss: 0.4839
Epoch 2/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8401 - loss: 0.4679 - val_accuracy: 0.8452 - val_loss: 0.4400
Epoch 3/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8474 - loss: 0.4400 - val_accuracy: 0.8488 - val_loss: 0.4338
Epoch 4/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8544 - loss: 0.4252 - val_accuracy: 0.8517 - val_loss: 0.4306
Epoch 5/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8569 - loss: 0.4159 - val_accuracy: 0.8532 - val_loss: 0.4250


In [5]:
test_loss, test_acc = model.evaluate(x_test, y_test, verbose=0)
print(f"✅ Logistic Regression 테스트 정확도: {test_acc:.4f}")


✅ Logistic Regression 테스트 정확도: 0.8398


## ✅ 정리
- ANN (chapter01): Dense(128, relu) + Dense(10, softmax) → 정확도 약 0.85~0.88
- Logistic Regression: Dense(10, softmax)만 사용 → 정확도 약 0.82~0.84

👉 로지스틱 회귀는 **기본 baseline 모델**로 ANN과 비교할 기준을 제공합니다.  
다음 챕터: **ANN 기본 구조 확장 & 성능 비교**
